Fruzsi txtfile

rajfruzsi opened the pull request:
Writing to and reading from txtfile feature created.

rajfruzsi committed:
added functions to load and save to txt, modified main to take exit as exitcommand and save and load from file

rajfruzsi committed:
corrected loading from file, added string splitter function

rajfruzsi committed:
removed unused include

removed algorithm

hegyhati commented on cmd.cpp:
```diff
@@ -1,39 +1,44 @@
-	for (unsigned int i = 0; i < this->system.size(); i++) {
-		if (currentFolder == this->system[i].folder && dirName == this->system[i].subfolder) {
-			found = true;
+	if (this->checkIfNameIsValid(dirName)) cout << "Name cannot contain special characters" << endl;
```
Szoval ha valid, akkor nem fog mukodni?

hegyhati commented on cmd.cpp:
```diff
@@ -1,39 +1,44 @@
+	if (this->checkIfNameIsValid(dirName)) cout << "Name cannot contain special characters" << endl;
+	else {
+		for (unsigned int i = 0; i < this->system.size(); i++) {
+			if (currentFolder == this->system[i].folder && dirName == this->system[i].subfolder) {
```
Ez nekem nagyon fura. itt a folder helyett nem valami "parent" jellegu dolog kellene, subfolder helyett meg name? a system[i] valami folder/subfolder parokat tarol?

hegyhati replied:
Jah, es ha igen, akkor mi van akkor, ha ket kulonbozo helyen is van "alma" konyvtaram?

rajfruzsi replied:
A system egy Pair-eket tároló vektor, és a Pairben folder és subfolder van. 
Lehetséges, hogy legyen két külön helyen is ,,alma" könyvtár.

hegyhati replied:
Az oke, es akkor ha az "alma"-ban akarok letrehozni egy "korte"-t, akkor honnet fogja tudni, melyikbe csinalja? MEg ennel bonyolultabb utkozesek is elerhetok nyilvan.

szilidaniel replied:
A System osztály párokat tárol. Egy mappa nevét és az abszolút elérési utat a mappához (subfolder nevet kicserélhetnénk Path szóra is igazából, de ezzel a névvel kezdtük szóval maradt). Egy CurrentDirectoryval követjük melyik mappában vagyunk. Ez nem a mappa neve, hanem az egész út a mappához így kezelni tudjuk az egyforma neveket. A parancsok mindig a CurrentDirectoryban hajtódnak végre és cd parancs esetén ezt módosítjuk nyilván.

hegyhati replied:
Akkor refaktoraljatok a kodot, es nevezzetek at normalis nevekre, amik tenyleg azt jelentik, amit tartalmaznak.

hegyhati commented on cmd.cpp:
```diff
@@ -1,39 +1,44 @@
+				this->addPairToVector(currentFolder, dirName);
+			}
+			else {
+				cout << "Invalid foldername" << endl;
```
szoval a `..` az egyetlen invalid foldername?

rajfruzsi replied:
True, ellenőrzö függvényt raktam ide, ami ellenőrzi. Az elnevezése nem volt a legmegfelelőbb, azt is javítottam.

hegyhati commented on cmd.cpp:
```diff
@@ -218,4 +224,65 @@ void Dictionary::touch(string fileName, string currentFolder) {
+Pair Dictionary::splitStringFirstSlash(string line) {
+	Pair p;
+	string path = line;
+	int cut = line.find_last_of("/");
```
szoval FirstSlash, erre find_last_of ? :-) 
De amugyis erre egy kulon fuggvenyt letrehozni kicsit overkill. ITt kb 2 sor az, ami a darabolassal megy el, a tobbi az, hogy valahogy belepakold ezt. De ha mindenkeppen ki van szedve, akkor lehetne mar eleve ennek a Pair-nek egy ilyen string alapu konstruktora, meg egy "toString" fuggvenye. Es akkor ez a logika relative kozel lenne egymashoz lekodolva. De ez csak otlet.

rajfruzsi replied:
True, nem a legjobb az elnevezés (megint), azt javítottam, és marad egyelőre ebben a formában, mert átláthatóbb. Az ötlet igazából nagyon jó, és azt akkor majd az egész kódra kéne alkalmaznunk, mert máshol is fölösleges ismétlődéseket spórolhatnánk vele. Ezt megvalósítom akkor.

rajfruzsi replied:
Konstruktor done. Todo, hogy a structot külön cpp-be kivegyük.

hegyhati replied:
Olvasgatom, amiket irsz, ha kerdes van, reagalok. Ha nincs kerdes, akkor ha nem irok, akkor ack-ztam. Arra kerlek, hogyha review ready ujra, akkor explicit irj a PR-hoz egy kommentet, amiben taggelsz, hogy nezzem meg ujra.

hegyhati replied:
Jah, es amit kesobbi PR-re akarsz elhalasztani, arrol mondjuk erdemes csinalni issue-t, hogy ne felejtsd el. Ha itt akarod megcsinalni, akkor tudsz itt ilyen todo list szeruseget is csinalni.

hegyhati reviewed (changes requested):
Egy csomo minden mas is van benne, nemcsak a fajlba kiiras, es vetodott fel kerdes. Plusz tessek a CI-ban tesztelni ezt a kimentes, betoltes dolgot is.

rajfruzsi committed:
corrected suggestions - renamed functions

rajfruzsi committed:
added default constructor, and parameterized too

rajfruzsi committed:
taken constructor into h and cpp

rajfruzsi committed:
Merge branch 'master' into fruzsi_txtfile

hegyhati commented on cmd.cpp:
```diff
@@ -1,11 +1,22 @@
+#include <fstream>
 using namespace std;
 
+Pair::Pair() {
```
van egyebkent olyan, hogy `std::pair`, de itt van letjogosultsaga a sajat osztalynak, mondjuk nem feltetlenul a `Pair` a legtalalobb nev :-)

szilidaniel replied:
Ez csak egy elnevezés. Nem az std::pairt akartuk használni.

hegyhati replied:
Ahogy irtam, itt van letjogosultsaga, hogy csinaltok sajat strukturat, es az adattagokat ertelmes (lasd masik thread) nevekkel el tudjatok erni a first/second helyett. De, ettol fuggetlenul ennek talan lehetne valami talalobb nevet adni annal, hogy `Pair`.  Ez tul altalanos nev, amivel ket baj van:
1) Ha ranez valaki a kodra, akkor fogalma sincs, mi ez.
2) Ha barhol egy kesobb berantott lib-ben van szinten egy Pair, akkor gyonyoruen ossze fog utkozni vele. (az ilyenek elkerulesere valok a namespace-ek, ugye)

hegyhati commented on cmd.cpp:
```diff
@@ -18,11 +29,8 @@ void Dictionary::mkdir(string dirName, string currentFolder) {
-			p.folder = currentFolder;
-			p.subfolder = dirName;
-			this->system.push_back(p);
+		if (!(this->isNotValid(dirName))) {
```
de szep a duplatagadas ;-) Lehet erdemesebb lenne az `if(isNotValid(dn)) cout<< ... else ... `, dee mar csak apro kodolvashatosagi dolog.

rajfruzsi replied:
Igaz :D Javítva, again.

hegyhati commented on cmd.cpp:
```diff
@@ -31,9 +39,11 @@ void Dictionary::mkdir(string dirName, string currentFolder) {
 	for (unsigned int i = 0; i < this->system.size(); i++) {
 		if (currentFolder == this->system[i].folder) {
-			cout << this->system[i].subfolder << endl;
+			if (this->system[i].subfolder != "root")
```
egy subfolder hogy lehetne root? vagy az inkabb parent folder?

szilidaniel replied:
subfolder =parent

hegyhati replied:
Again, refaktoraljatok, hogy ne legyen felrevezeto a kod.

hegyhati commented on cmd.cpp:
```diff
@@ -219,6 +229,73 @@ void Dictionary::touch(string fileName, string currentFolder) {
+
+void Dictionary::addPairToVector(string folder, string subfolder) {
+	Pair p(folder, subfolder);
+	this->system.push_back(p);
```
Ez akar igazan lehetne egy sorba is, es akkor kevesebb copy constructor megy vegbe, de mondjuk az optimalizalo ugy is kiszedi.

rajfruzsi replied:
True. Javítva.

hegyhati commented on cmd.h:
```diff
@@ -20,6 +22,7 @@ class Dictionary {
 private:
 	vector<Pair> system;
 	vector<FileDescriptor> fileDescriptorVector;
+	vector<string> commands = { "mkdir","exit","ls","cd","rm","touch" };
```
az oke, hogy ezt hasznalod arra, hogy megallapitsd, hogy egy command valid-e, de utana ha jol sejtem, arra nem tudod hasznalni, hogy melyik commandhoz melyik fuggveny legyen meghivva. Innentol kezdve olyan nagyon sok ertelme nincs, mert gondolom ott van egy jo nagy if, aminek a vegere egy `else not valid`-dal ez megoldhato lenne. Illetve ha most van egy uj command, akkor ott is bele kell tenni egy uj else if-et, meg itt is ki kell egesziteni. Tehat ket helyen kell turkalni a kodot. Jol sejtem?

hegyhati commented on main.cpp:
```diff
@@ -20,29 +10,37 @@ vector<string> split(string path) {
+		cout << "Not valid name. Please add a valid file name to load filesystem from. " << endl;
+		cin >> fsname;
+		if (fsname == "exit") break;
+	}
```
Na, egyreszt ket kiiratasod isvan, az egyik ker egy filenevet, a masik meg mar parancsokat var. 
De aztan a fajlnevet olvasod be eloszor, amirol leteszteled, hogy valid command-e? (minek?!)
Illetve ez typic helye egy `do while` ciklusnak, amikor ilyet akarunk :-)

rajfruzsi replied:
A második kiiratás az igazából egy ilyen knowhow hogy hogyan lehet exitelni, nem vár még ott parancsot. Az első kiiratás pedig arról szól, hogy adjon meg valid fájlnevet. A validcommanddel azt szűröm ki, hogy az input ne lehessen valami parancs, mert először fájlnév kell. De a szöveg ezt nem feltétlen impikálja, szóval javítottam.

hegyhati replied:
Oke, igy ertem, meg annyi, hogy talan logikusabb lenne az a ket kiiratas akkor forditott sorrendben.De ez mar csak nitpicking.

hegyhati commented on main.cpp:
```diff
@@ -20,29 +10,37 @@ vector<string> split(string path) {
 		cout << autotext;
 		cin >> parancs;
-		if (!validcommand(parancs)) {
+		if (parancs == "exit") break;
```
na itt ez az if kaosz ami lentebb van mar atlathatatlan, tessek rendbeszedni :-) 
Egyreszt a parancsok kozotti szetvalasztas legyen egy levelen, masreszt ilyen /-enkenti parzolgatast ne a main-be tegyunk mar bele.

rajfruzsi replied:
Javítottam, rerequesteltem, a parzolast száműztem functionbe. :)

hegyhati reviewed (changes requested):
Par stilisztikai dolog, meg egy-ket komolyabb.

rajfruzsi committed:
updated exit command

changed q to exit

rajfruzsi committed:
corrected 2 line function

corrected constructor instantiating in function

rajfruzsi committed:
corrected filename check

Edited cout text to let user know about entering command name instead of a filename.

rajfruzsi committed:
corrected function name

corrected bool function isNotValid to IsValid name for better understanding of code.

rajfruzsi committed:
corrected function name

corrected bool function isNotValid to IsValid name for better understanding of code.

rajfruzsi committed:
corrected instantiating

corrected instantiating of pair

rajfruzsi committed:
corrected if else mess, took out some parts to individual functions

rajfruzsi committed:
merge conflict solved

rajfruzsi committed:
Merge branch 'master' into fruzsi_txtfile

hegyhati commented on cmd.cpp:
```diff
@@ -13,11 +13,18 @@ Pair::Pair(string folder, string subfolder) {
 }
 
 void Dictionary::mkdir(string dirName, string currentFolder) {
+	/**@brief Mappa létrehozása
+	*@param dirName egy mappanevet kell megadni a függvénynek, mégpedig amilyen néven létre szeretnénk hozni.
+	*@param currentFolder át kell adni az aktuális tartózkodási helyet a mapparendszeren belül és ott kerül létrehozásra a mappa.
+	*/
```
egy branch egy dolgot csinaljon :-)

hegyhati commented on cmd.cpp:
```diff
@@ -13,11 +13,18 @@ Pair::Pair(string folder, string subfolder) {
-				found = true;
+			found = true;
 		}
+		/**@brief Mappa ellenõrzése
+		*@return Igaz értéket ad vissza, ha a mappa létezik.
+		*/
```
uch, ide igy nem.

hegyhati reviewed (approved):
Az a Pair dolog tovabbra is nagyon csunya, meg par aprosag, de huzodiuk mar egy ideje, menjen most be.
